components/org.wso2.carbon.event.outbound.adapter.websubhub/src/main/java/org/wso2/carbon/event/outbound/adapter/websubhub/ClientManager.java

```diff
@@ -0,0 +1,173 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.event.outbound.adapter.websubhub;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.conn.ssl.DefaultHostnameVerifier;
+import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder;
+import org.apache.http.impl.nio.client.HttpAsyncClients;
+import org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager;
+import org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor;
+import org.apache.http.nio.reactor.ConnectingIOReactor;
+import org.apache.http.ssl.SSLContexts;
+import org.wso2.carbon.event.outbound.adapter.websubhub.internal.WebSubHubEventAdapterDataHolder;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+
+import java.io.IOException;
+import java.security.KeyManagementException;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.net.ssl.SSLContext;
+
+import static org.wso2.carbon.event.outbound.adapter.websubhub.util.WebSubHubEventAdapterConstants.CONNECTION_POOL_MAX_CONNECTIONS;
+import static org.wso2.carbon.event.outbound.adapter.websubhub.util.WebSubHubEventAdapterConstants.CONNECTION_POOL_MAX_CONNECTIONS_PER_ROUTE;
+
+/**
+ * Class to retrieve the HTTP Clients.
+ */
+public class ClientManager {
+
+    private static final Log LOG = LogFactory.getLog(ClientManager.class);
+    private static final Map<Integer, CloseableHttpAsyncClient> clientMap = new HashMap<>();
+    private static final int HTTP_CONNECTION_TIMEOUT = 300;
+    private static final int HTTP_READ_TIMEOUT = 300;
+    private static final int HTTP_CONNECTION_REQUEST_TIMEOUT = 300;
+    private static final int DEFAULT_MAX_CONNECTIONS = 20;
+    private final PoolingNHttpClientConnectionManager poolingHttpClientConnectionManager;
+
+    /**
+     * Creates a client manager.
+     *
+     * @throws IOException on errors while creating the pooling connection manager.
+     */
+    public ClientManager() throws IOException {
+
+        poolingHttpClientConnectionManager = createPoolingConnectionManager();
+    }
+
+    /**
+     * Get HTTP client properly configured with tenant configurations.
+     *
+     * @param tenantDomain Tenant domain of the service provider.
+     * @return HttpClient.
+     */
+    public CloseableHttpAsyncClient getClient(String tenantDomain) throws IOException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        CloseableHttpAsyncClient client = clientMap.get(tenantId);
```

Comment:
Can we use a single client instead of keeping a tenant-wise client map. There was a suggestion to improve the Client Manager in previous implementation for this.